src/net/sourceforge/kolmafia/textui/Parser.java

```diff
@@ -4461,6 +4475,44 @@ public static Location mergeLocations(final Command start, final Command end) {
 
   // **************** Parse errors *****************
 
+  public class AshDiagnostic {
+    final Location location;
+    final DiagnosticSeverity severity;
+    final String message;
+    final List<String> additionalMessages;
+
+    private AshDiagnostic(
+        final Location location,
+        final DiagnosticSeverity severity,
+        final String message,
+        final String... additionalMessages) {
+      this.location = location;
+      this.severity = severity;
+      this.message = message;
+      this.additionalMessages = new ArrayList<>();
+      for (final String additionalMessage : additionalMessages) {
+        this.additionalMessages.add(additionalMessage);
+      }
+    }
+
+    public String toString() {
+      StringBuilder result = new StringBuilder();
+
+      result.append(this.message);
+      result.append(" (");
+
+      result.append(Parser.getFileAndRange(Parser.this.shortFileName, this.location.getRange()));
```

Comment:
This means that if you toString an ashdaognostic after the fact, you may get the wrong filename because the parser is looking at a different file

Reply:
This is actually the same as line 4802
```java
  private ScriptException parseException(Location location, final String message) {
    if (location == null) {
      location = this.makeZeroWidthLocation();
    }

    return new ScriptException(
        message + " (" + Parser.getFileAndRange(this.shortFileName, location.getRange()) + ")");
  }
```

This is *intentional*.
If "the parser is looking at a different file", then it means that the AshDiagnostic was put in the wrong Parser.
A Parser should only report errors happening in the file it parsed.

I could add a check that the submitted location's URI matches shortFileName, but I'm worried about false positives, due to not knowing how File.getUri() works with every OS...

Reply:
Someone could take a copy of the AshDiagnostic produced and expect it to be valid later on in execution though. You should change this to ensure that we don't need to rely on your having made sure it never happens.

Reply:
What exactly do you mean by taking a copy of it?
You mean adding it to a Parser's `diagnostics` field?

Reply:
In the future code could be written that holds an AshDiagnostic and does something with it later.

Reply:
That was... shamefully easier than expected (o///o)